src/cdi_CPEloss/Tabular_CP_Eloss.hh

```diff
@@ -0,0 +1,115 @@
+ *        in a file.                
+ *                                                                               
+ */
+//===========================================================================//
+
+class Tabular_CP_Eloss : public rtt_cdi::CPEloss {
+public:
+  typedef std::vector<double> sf_double;
+  typedef stdex::extents<stdex::dynamic_extent,
+                         std::experimental::dynamic_extent,
+                         std::experimental::dynamic_extent>
+      dynamic_extents_3;
+
+private:
+  const std::string filename;
+  std::ifstream file;
+
+  rtt_units::UnitSystem us;
+  rtt_units::PhysicalConstants pc_cgs;
+
+  uint32_t n_energy;          //!< Number of gridpoints in projectile energy
+  uint32_t n_density;         //!< Number of gridpoints in target density
+  uint32_t n_temperature;     //!< Number of gridpoints in target temperature
+  double d_log_energy;        //!< Log spacing of projectile energy gridpoints
+  double d_log_density;       //!< Log spacing of target density gridpoints
+  double d_log_temperature;   //!< Log spacing of target temperature gridpoints
+  double min_log_energy;      //!< Log of minimum projectile energy
+  double min_log_density;     //!< Log of minimum target density
+  double min_log_temperature; //!< Log of minimum target temperature
+  double min_energy;          //!< Minimum target energy
+  double max_energy;          //!< Maximum target energy
+  double min_density;         //!< Minimum target density
+  double max_density;         //!< Maximum target density
+  double min_temperature;     //!< Minimum target temperature
+  double max_temperature;     //!< Maximum target temperature
+  sf_double energies;         //!< Vector of energy gridpoints
+  sf_double densities;        //!< Vector of density gridpoints
+  sf_double temperatures;     //!< Vector of temperature gridpoints
+  // Note that after unit conversions, *_energy is really *_speed
+
+  // Storage for tabulated data
+  stdex::basic_mdspan<double, dynamic_extents_3, std::experimental::layout_left>
+      stopping_data;
+
+  // Utility for skipping lines
+  void skip_lines(uint32_t nlines);
+
+  // Utility for reading a line of an eloss file and as a vector of strings
+  std::vector<std::string> const read_line();
+
+public:
+  // Constructor
+  Tabular_CP_Eloss(std::string filename_in, rtt_cdi::CParticle target_in,
+                   rtt_cdi::CParticle projectile_in);
+
+  // >>> ACCESSORS
+
+  double getEloss(const double temperature, const double density,
+                  const double v0) const;
+
+  bool data_in_tabular_form() const { return true; }
+
+  inline std::string getDataFilename() const { return filename; }
+
+  sf_double getTemperatureGrid() const { return temperatures; }
+
+  sf_double getDensityGrid() const { return densities; }
+
+  sf_double getEnergyGrid() const { return energies; }
+
+  size_t getNumTemperatures() const { return n_temperature; }
```

Comment:
Instead of saving `n_temperature`, does it make sense to return `temperatures.size()`?  This might help avoid the need to manually update the value of `n_temperature`.

Reply:
I worry that this will make it harder to detect inconsistencies in bad data files, which provide `n_temperature` directly. Right now, we need to read in and store `n_temperature` at some point anyway, so I'm inclined to just leave this part of the code as-is.

Reply:
okay.